### PR TITLE
Add support to reindex the database and remove events.

### PR DESCRIPTION
**Pull request by poljar**
This is a bit of a big one and it essentially implements two features.

While developing support to remove events from the database/index it has come to my attention that documents can only be deleted by a field if the field is indexed.
Tantivy has no way to find a document by a field if the field isn't indexed (doh).

This in turn requires a schema change on our part since we really need to delete events by event id, which is unique and guarantees that we're only going to remove this one intended event. 

Our current Tantivy schema is lacking some more definitions to support the following use cases:
* As mentioned, support to remove redacted events https://github.com/vector-im/riot-web/issues/12386
* Search by sender/author https://github.com/matrix-org/matrix-react-sdk/pull/4156
* Search by date https://github.com/vector-im/riot-web/issues/4752

While this PR doesn't implement all of the things that are needed to support those features it does make the required Tantivy schema changes and bumps the database version.

To allow breaking schema changes without the need to delete the whole database and re-download the complete room history we introduce a `RecoveryDatabase`, this database acts as a read only database that allows the index to be deleted while keeping the SQLite database in tact.

The `RecoveryDatabase` can then proceed to create a new index and fetch events out of the database for them to be re-added to the index.

As already mentioned, support to remove events from the database and index is included here as well. Removing an event requires a Tantivy commit, so they need to go through the flow control logic the same way as newly added events.

Some things to consider during review:
* Do we want all the non-implemented features as well
* Since re-indexing requires events to be de-serialized and `neon-serde` can't be used for this; do we want to make this feature optional since it requires `serde_json`.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -51,6 +54,9 @@ pub enum Error {
     /// Error signaling that sqlcipher support is missing.
     #[fail(display = "Sqlcipher error: {}", _0)]
     SqlCipherError(String),
+    /// Error indicating that the index needs to be rebuild.
```
Comment (jryans) [inline src/error.rs]:
```suggestion
    /// Error indicating that the index needs to be rebuilt.
```

File: src/error.rs [hunk 2]
```diff
@@ -51,6 +54,9 @@ pub enum Error {
...
+    #[fail(display = "Error opening the database, the index needs to be rebuild.")]
```
Comment (jryans) [inline src/error.rs]:
```suggestion
    #[fail(display = "Error opening the database, the index needs to be rebuilt.")]
```

File: src/database/static_methods.rs [hunk 3]
```diff
@@ -62,6 +62,62 @@ impl Database {
...
+        connection: &mut rusqlite::Connection,
+        index_writer: &mut IndexWriter,
+        event_id: EventId,
+        undeleted_events: &mut Vec<EventId>,
+    ) -> Result<bool> {
+        let transaction = connection.transaction()?;
+
+        transaction.execute("DELETE from events WHERE event_id == ?1", &[&event_id])?;
+        transaction.execute(
+            "INSERT OR IGNORE INTO undeleted_events (event_id) VALUES (?1)",
```
Comment (jryans) [inline src/database/static_methods.rs]: Hmm, does `undeleted_events` mean "events to be deleted"? If so, "undeleted" is confusing to me as I would assumed it means "restored" or "brought back from the trash".

If my assumption is correct, maybe use `pending_deletion_events` or something?

File: seshat-node/lib/index.js [hunk 4]
```diff
@@ -393,4 +445,64 @@ class Seshat extends seshat.Seshat {
...
+ *
+ * @constructor
+ *
+ * @example
+ * // open a Seshat recovery database in the given folder
+ * let recovery = new SeshatRecovery("/home/example/database_dir");
+ * // reindex the database
+ * await recovery.reindex();
+ */
+class SeshatRecovery extends seshat.SeshatRecovery {
```
Comment (jryans) [inline seshat-node/lib/index.js]: I might have feedback after seeing how you plan to handle the recovery path in Riot, but at an abstract level it seems okay.

File: src/database/static_methods.rs (hunk 3 above)
Comment (poljar) [reply src/database/static_methods.rs]: It is events to be deleted.

File: seshat-node/lib/index.js (hunk 4 above)
Comment (poljar) [reply seshat-node/lib/index.js]: This commit implements support for it in Riot: https://github.com/vector-im/riot-web/commit/921d734ef65de26f5a58394beee4ac001ee87085.

I didn't notice any delays at startup, the thing seems to be reasonably quick but do note that this was using a 3k events large database.

**Comment by poljar** [comment]
> I would think all clients would want to be able to support re-indexing, so I would say it's fine to be required until there's a problem... I assume it works fine with Riot?

It does work fine with Riot, the only thing that worries me is a bit of code/dependency bloat.

File: src/database/static_methods.rs (hunk 3 above)
Comment (ara4n) [reply src/database/static_methods.rs]: +1 tp `pending_deletion_events`

**Pull request merged**